fix: handle mismatched types in newOwners/new_owners to prevent NFT loss

SoarinSkySagar opened the pull request:
## Description
Closes issue #227.

## Tasks:
- [x] Modified the `Request` struct in `Protocol.sol` to have `address` type for `newOwners`
- [x] Created a new function `cairoAddressArrayDeserialize` in `Cairo.sol` to deserialize `newOwners` array, with proper comments above the function, in consistence with the rest of the file.
- [x] Used the newly created `cairoAddressArrayDeserialize` function to deserialize the `address` array in `Protocol.sol`.
- [x] Modified the unit tests for `Protocol.sol` in `Protocol.t.sol` so that it is being tested against the proper `newOwners` type.
- [x] Created the unit test function `test_cairoAddressArrayDeserialize` in `Cairo.t.sol` for the function `cairoAddressArrayDeserialize`.

*This PR was created as part of OnlyDust's ODHack 8.0.*

SoarinSkySagar commented:
Hi @ptisserand, I have completed all the required tasks and ensured good coding practices in the commit (Issue #227), please review and merge!

ptisserand commented:
Hi, thanks for your contribution but there is a compilation issue when trying to run tests.
Could you please fix it?

SoarinSkySagar commented:
hi @ptisserand, yesterday when I ran the test it was compiling, can you please share the screenshot?

ptisserand commented:
> hi @ptisserand, yesterday when I ran the test it was compiling, can you please share the screenshot of the error so I can look into it?

Ok, maybe your forgot to commit & push, here is the errors:
```
--> src/TestMessaging.sol

Error (7576): Undeclared identifier.
   --> test/Cairo.t.sol:125:9:
    |
125 |         buf1[0] = 0;
    |         ^^^^

Error (7576): Undeclared identifier.
   --> test/Cairo.t.sol:126:93:
    |
126 |         (address[] memory result1, uint256 newOffset1) = Cairo.cairoAddressArrayDeserialize(buf1, 0);
    |                                                                                             ^^^^

Error (7576): Undeclared identifier.
   --> test/Cairo.t.sol:131:9:
    |
131 |         buf2[0] = 2; 
    |         ^^^^

Error (7576): Undeclared identifier.
   --> test/Cairo.t.sol:132:9:
    |
132 |         buf2[1] = uint160(address(0x1234567890abcdef1234567890abcdef12345678));
    |         ^^^^

Error (7576): Undeclared identifier.
   --> test/Cairo.t.sol:133:9:
    |
133 |         buf2[2] = uint160(address(0xabcdefabcdefabcdefabcdefabcdefabcdefabcd));
    |         ^^^^

Error (7576): Undeclared identifier.
   --> test/Cairo.t.sol:135:93:
    |
135 |         (address[] memory result2, uint256 newOffset2) = Cairo.cairoAddressArrayDeserialize(buf2, 0);
    |                                                                                             ^^^^

Error (7576): Undeclared identifier.
   --> test/Cairo.t.sol:142:9:
    |
142 |         buf3[0] = 1; 
    |         ^^^^

Error (7576): Undeclared identifier.
   --> test/Cairo.t.sol:143:9:
    |
143 |         buf3[1] = uint160(address(0xdeadbeefdeadbeefdeadbeefdeadbeefdeadbeef));
    |         ^^^^

Error (7576): Undeclared identifier.
   --> test/Cairo.t.sol:145:93:
    |
145 |         (address[] memory result3, uint256 newOffset3) = Cairo.cairoAddressArrayDeserialize(buf3, 0);
    |                                                                                             ^^^^

Error: 
Compilation failed
```

SoarinSkySagar commented:
Hi @ptisserand, can you please check now? Earlier the forge commands were running on the main branch of the forked repo whereas I was working on a new branch (even though I changed the branch to the working one), hence no errors were generated. Had to re-clone the forked repo and the errors were there as you said. Here's the new checklist of tasks completed:
- [x] Created a new function `addressArraySerialize` and its corresponding unit test.
- [x] Fixed some type mismatch errors in the `test_cairoAddressArrayDeserialize` function.
- [x] Checked to see that all tests are passing.

### Screenshot of compilation
![image](https://github.com/user-attachments/assets/b1f6d37a-ef7e-401c-9554-dbb4c08d9174)

ptisserand commented on apps/blockchain/ethereum/src/sn/Cairo.sol:
```diff
@@ -313,6 +345,33 @@ library Cairo {
+     @param buf Buffer where data is serialized
+     @param offset Offset in 'buf' where deserialization must start
+
+     @return The offset increment applied to deserialize and the deserialized value.
+    */
+
+    function cairoAddressArrayDeserialize(
+        uint256[] memory buf,
+        uint256 offset
+    )
+        internal
+        pure
+        returns (address[] memory, uint256)
+    {
```
Please use same pattern as other functions 
```solidity
returns (uint256, address[] memory)
```

ptisserand commented on apps/blockchain/ethereum/test/Cairo.t.sol:
```diff
@@ -118,6 +118,72 @@ contract CairoTest is Test {
+        assertEq(result1.length, 0);
+        assertEq(newOffset1, 1); 
+
+        uint256;
```
Please remove this line

ptisserand commented on apps/blockchain/ethereum/test/Cairo.t.sol:
```diff
@@ -118,6 +118,72 @@ contract CairoTest is Test {
+        assertEq(result2[1], address(0xABcdEFABcdEFabcdEfAbCdefabcdeFABcDEFabCD));
+        assertEq(newOffset2, 3);
+
+        uint256;
```
Please remove this line

ptisserand commented on apps/blockchain/ethereum/test/Cairo.t.sol:
```diff
@@ -118,6 +118,72 @@ contract CairoTest is Test {
+        uint256[] memory buf1 = new uint256[](3);
+        uint256 offset1 = Cairo.addressArraySerialize(arr1, buf1, 0);
+
+        assertEq(buf1[0], 2); 
```
Could you please add a check on  `offset1` value

ptisserand commented on apps/blockchain/ethereum/test/Cairo.t.sol:
```diff
@@ -118,6 +118,72 @@ contract CairoTest is Test {
+        uint256[] memory buf2 = new uint256[](2);
+        uint256 offset2 = Cairo.addressArraySerialize(arr2, buf2, 0);
+
+        assertEq(buf2[0], 1);
```
Could you please add a check on `offset2` value

ptisserand commented on apps/blockchain/ethereum/test/Cairo.t.sol:
```diff
@@ -118,6 +118,72 @@ contract CairoTest is Test {
+
+        uint256 offset3 = Cairo.addressArraySerialize(arr3, buf3, 0);
+
+        assertEq(buf3[0], 0);
```
Could you please add a check on `offset3` value

ptisserand reviewed (changes requested):
Thanks for your PR, some minor fixes remaining and it should be good to been merged.

SoarinSkySagar commented:
Hi @ptisserand, hope its okay now and and merge-able :)

ptisserand reviewed (approved):
Thanks for your PR!
Let's get it merged
